src/app/src/test/kotlin/org/vaccineimpact/api/tests/controllers/DirectActionContextTests.kt

```diff
@@ -85,11 +89,132 @@ class DirectActionContextTests : MontaguTests()
         assertThat(model).isEqualTo(Model(1, listOf("x", "y"), Model(100, listOf("z"), null)))
     }
 
+    @Test
+    fun `throw BadRequest if getPart called on non multipart request`()
+    {
+        val profile = CommonProfile()
+
+        val mockRequest = mock<Request> {
+            on { raw() } doReturn mock<HttpServletRequest>()
+            on { contentType() } doReturn "text/plain"
+        }
+
+        val webContext = mock<SparkWebContext> {
+            on { getRequestAttribute(Pac4jConstants.USER_PROFILES) } doReturn profile
+            on { sparkRequest } doReturn mockRequest
+        }
+
+        val context = DirectActionContext(webContext)
+        assertThatThrownBy {
+            context.getPart("whatever")
+        }.isInstanceOf(BadRequest::class.java)
+                .hasMessageContaining("Trying to extract a part from multipart/form-data but this request is of type text/plain")
+    }
+
+    @Test
+    fun `throw BadRequest if getPart called on non existent part`()
+    {
+        val profile = CommonProfile()
+
+        val mockRequest = mock<Request> {
+            on { raw() } doReturn mock<HttpServletRequest>()
+            on { contentType() } doReturn "multipart/form-data"
+        }
+
+        val webContext = mock<SparkWebContext> {
+            on { getRequestAttribute(Pac4jConstants.USER_PROFILES) } doReturn profile
+            on { sparkRequest } doReturn mockRequest
+        }
+
+        val context = DirectActionContext(webContext)
+        assertThatThrownBy {
+            context.getPart("whatever")
+        }.isInstanceOf(BadRequest::class.java)
+                .hasMessageContaining("No value passed for required POST parameter 'whatever'")
+    }
+
+    @Test
+    fun `can get part`()
+    {
+        val profile = CommonProfile()
+
+        val mockServletRequest = mock<HttpServletRequest>() {
+            on { getPart("whatever") } doReturn FakePart()
+        }
+
+        val mockRequest = mock<Request> {
+            on { raw() } doReturn mockServletRequest
+            on { contentType() } doReturn "multipart/form-data"
+        }
+
+        val webContext = mock<SparkWebContext> {
+            on { getRequestAttribute(Pac4jConstants.USER_PROFILES) } doReturn profile
+            on { sparkRequest } doReturn mockRequest
+        }
+
+        val context = DirectActionContext(webContext)
+        assertThat(context.getPart("whatever")).isEqualTo("something")
+    }
+
     private fun mockWebContext(profile: CommonProfile): SparkWebContext
     {
         val webContext = mock<SparkWebContext> {
             on { getRequestAttribute(Pac4jConstants.USER_PROFILES) } doReturn profile
         }
         return webContext
     }
+
+    private class FakePart : Part
```

Comment:
I assume there's a reason why you can't just use a mock here?

Reply:
Huh, looks to me now like absolutely no reason! Not sure what I was thinking there...